// @portabletext/editor@2.14.2/lib/_chunks-cjs/selector.is-active-style.cjs
"use strict";
var types = require("@sanity/types"), selector_getSelectionText = require("./selector.get-selection-text.cjs"), util_isEmptyTextBlock = require("./util.is-empty-text-block.cjs"), schema = require("@portabletext/schema"), util_getTextBlockText = require("./util.get-text-block-text.cjs");
const getFocusInlineObject = (snapshot) => {
  const focusChild = selector_getSelectionText.getFocusChild(snapshot);
  return focusChild && !types.isPortableTextSpan(focusChild.node) ? {
    node: focusChild.node,
    path: focusChild.path
  } : void 0;
}, getSelectedBlocks = (snapshot) => {
  if (!snapshot.context.selection)
    return [];
  const selectedBlocks = [], startPoint = util_getTextBlockText.getSelectionStartPoint(snapshot.context.selection), endPoint = util_getTextBlockText.getSelectionEndPoint(snapshot.context.selection), startKey = util_getTextBlockText.getBlockKeyFromSelectionPoint(startPoint), endKey = util_getTextBlockText.getBlockKeyFromSelectionPoint(endPoint);
  if (!startKey || !endKey)
    return selectedBlocks;
  const startBlockIndex = snapshot.blockIndexMap.get(startKey), endBlockIndex = snapshot.blockIndexMap.get(endKey);
  if (startBlockIndex === void 0 || endBlockIndex === void 0)
    return selectedBlocks;
  const slicedValue = snapshot.context.value.slice(startBlockIndex, endBlockIndex + 1);
  for (const block of slicedValue) {
    if (block._key === startKey) {
      if (selectedBlocks.push({
        node: block,
        path: [{
          _key: block._key
        }]
      }), startKey === endKey)
        break;
      continue;
    }
    if (block._key === endKey) {
      selectedBlocks.push({
        node: block,
        path: [{
          _key: block._key
        }]
      });
      break;
    }
    selectedBlocks.length > 0 && selectedBlocks.push({
      node: block,
      path: [{
        _key: block._key
      }]
    });
  }
  return selectedBlocks;
}, getSelectionEndBlock = (snapshot) => {
  const endPoint = util_getTextBlockText.getSelectionEndPoint(snapshot.context.selection);
  if (endPoint)
    return selector_getSelectionText.getFocusBlock({
      ...snapshot,
      context: {
        ...snapshot.context,
        selection: {
          anchor: endPoint,
          focus: endPoint
        }
      }
    });
}, getSelectionStartBlock = (snapshot) => {
  const startPoint = util_getTextBlockText.getSelectionStartPoint(snapshot.context.selection);
  if (startPoint)
    return selector_getSelectionText.getFocusBlock({
      ...snapshot,
      context: {
        ...snapshot.context,
        selection: {
          anchor: startPoint,
          focus: startPoint
        }
      }
    });
}, getSelectionEndPoint = (snapshot) => {
  if (snapshot.context.selection)
    return snapshot.context.selection.backward ? snapshot.context.selection.anchor : snapshot.context.selection.focus;
};
function isPointAfterSelection(point) {
  return (snapshot) => {
    if (!snapshot.context.selection)
      return !1;
    const endPoint = util_getTextBlockText.getSelectionEndPoint(snapshot.context.selection), endBlockKey = util_getTextBlockText.getBlockKeyFromSelectionPoint(endPoint), endChildKey = util_getTextBlockText.getChildKeyFromSelectionPoint(endPoint), pointBlockKey = util_getTextBlockText.getBlockKeyFromSelectionPoint(point), pointChildKey = util_getTextBlockText.getChildKeyFromSelectionPoint(point);
    if (!pointBlockKey || !endBlockKey)
      return !1;
    const pointBlockIndex = snapshot.blockIndexMap.get(pointBlockKey), endBlockIndex = snapshot.blockIndexMap.get(endBlockKey);
    if (pointBlockIndex === void 0 || endBlockIndex === void 0)
      return !1;
    if (pointBlockIndex > endBlockIndex)
      return !0;
    if (pointBlockIndex < endBlockIndex)
      return !1;
    const pointBlock = snapshot.context.value.at(pointBlockIndex);
    if (!pointBlock || !schema.isTextBlock(snapshot.context, pointBlock))
      return !1;
    let pointChildIndex, endChildIndex, childIndex = -1;
    for (const child of pointBlock.children) {
      if (childIndex++, child._key === pointChildKey && child._key === endChildKey)
        return point.offset > endPoint.offset;
      if (child._key === pointChildKey && (pointChildIndex = childIndex), child._key === endChildKey && (endChildIndex = childIndex), pointChildIndex !== void 0 && endChildIndex !== void 0)
        break;
    }
    return pointChildIndex === void 0 || endChildIndex === void 0 ? !1 : pointChildIndex > endChildIndex;
  };
}
function isPointBeforeSelection(point) {
  return (snapshot) => {
    if (!snapshot.context.selection)
      return !1;
    const startPoint = util_getTextBlockText.getSelectionStartPoint(snapshot.context.selection), startBlockKey = util_getTextBlockText.getBlockKeyFromSelectionPoint(startPoint), startChildKey = util_getTextBlockText.getChildKeyFromSelectionPoint(startPoint), pointBlockKey = util_getTextBlockText.getBlockKeyFromSelectionPoint(point), pointChildKey = util_getTextBlockText.getChildKeyFromSelectionPoint(point);
    if (!pointBlockKey || !startBlockKey)
      return !1;
    const startBlockIndex = snapshot.blockIndexMap.get(startBlockKey), pointBlockIndex = snapshot.blockIndexMap.get(pointBlockKey);
    if (startBlockIndex === void 0 || pointBlockIndex === void 0)
      return !1;
    if (pointBlockIndex < startBlockIndex)
      return !0;
    if (pointBlockIndex > startBlockIndex)
      return !1;
    const pointBlock = snapshot.context.value.at(pointBlockIndex);
    if (!pointBlock || !schema.isTextBlock(snapshot.context, pointBlock))
      return !1;
    let pointChildIndex, startChildIndex, childIndex = -1;
    for (const child of pointBlock.children) {
      if (childIndex++, child._key === pointChildKey && child._key === startChildKey)
        return point.offset < startPoint.offset;
      if (child._key === pointChildKey && (pointChildIndex = childIndex), child._key === startChildKey && (startChildIndex = childIndex), pointChildIndex !== void 0 && startChildIndex !== void 0)
        break;
    }
    return pointChildIndex === void 0 || startChildIndex === void 0 ? !1 : pointChildIndex < startChildIndex;
  };
}
function isOverlappingSelection(selection) {
  return (snapshot) => {
    if (!selection || !snapshot.context.selection)
      return !1;
    const selectionStartPoint = selector_getSelectionText.getSelectionStartPoint({
      context: {
        ...snapshot.context,
        selection
      }
    }), selectionEndPoint = getSelectionEndPoint({
      context: {
        ...snapshot.context,
        selection
      }
    }), originalSelectionStartPoint = selector_getSelectionText.getSelectionStartPoint(snapshot), originalSelectionEndPoint = getSelectionEndPoint(snapshot);
    if (!selectionStartPoint || !selectionEndPoint || !originalSelectionStartPoint || !originalSelectionEndPoint)
      return !1;
    const startPointEqualToOriginalStartPoint = util_isEmptyTextBlock.isEqualSelectionPoints(selectionStartPoint, originalSelectionStartPoint), endPointEqualToOriginalEndPoint = util_isEmptyTextBlock.isEqualSelectionPoints(selectionEndPoint, originalSelectionEndPoint);
    if (startPointEqualToOriginalStartPoint && endPointEqualToOriginalEndPoint)
      return !0;
    const startPointBeforeSelection = isPointBeforeSelection(selectionStartPoint)(snapshot), startPointAfterSelection = isPointAfterSelection(selectionStartPoint)(snapshot), endPointBeforeSelection = isPointBeforeSelection(selectionEndPoint)(snapshot), endPointAfterSelection = isPointAfterSelection(selectionEndPoint)(snapshot), originalStartPointBeforeStartPoint = isPointBeforeSelection(originalSelectionStartPoint)({
      ...snapshot,
      context: {
        ...snapshot.context,
        selection: {
          anchor: selectionStartPoint,
          focus: selectionStartPoint
        }
      }
    }), originalStartPointAfterStartPoint = isPointAfterSelection(originalSelectionStartPoint)({
      ...snapshot,
      context: {
        ...snapshot.context,
        selection: {
          anchor: selectionStartPoint,
          focus: selectionStartPoint
        }
      }
    }), originalEndPointBeforeEndPoint = isPointBeforeSelection(originalSelectionEndPoint)({
      ...snapshot,
      context: {
        ...snapshot.context,
        selection: {
          anchor: selectionEndPoint,
          focus: selectionEndPoint
        }
      }
    }), originalEndPointAfterEndPoint = isPointAfterSelection(originalSelectionEndPoint)({
      ...snapshot,
      context: {
        ...snapshot.context,
        selection: {
          anchor: selectionEndPoint,
          focus: selectionEndPoint
        }
      }
    }), startPointEqualToOriginalEndPoint = util_isEmptyTextBlock.isEqualSelectionPoints(selectionStartPoint, originalSelectionEndPoint), endPointEqualToOriginalStartPoint = util_isEmptyTextBlock.isEqualSelectionPoints(selectionEndPoint, originalSelectionStartPoint);
    return !endPointEqualToOriginalStartPoint && !startPointEqualToOriginalEndPoint && !originalStartPointBeforeStartPoint && !originalStartPointAfterStartPoint && !originalEndPointBeforeEndPoint && !originalEndPointAfterEndPoint || endPointBeforeSelection && !endPointEqualToOriginalStartPoint || startPointAfterSelection && !startPointEqualToOriginalEndPoint ? !1 : !originalStartPointBeforeStartPoint && originalStartPointAfterStartPoint && !originalEndPointBeforeEndPoint && originalEndPointAfterEndPoint ? !endPointEqualToOriginalStartPoint : originalStartPointBeforeStartPoint && !originalStartPointAfterStartPoint && originalEndPointBeforeEndPoint && !originalEndPointAfterEndPoint ? !startPointEqualToOriginalEndPoint : !startPointAfterSelection || !startPointBeforeSelection || !endPointAfterSelection || !endPointBeforeSelection;
  };
}
const isSelectingEntireBlocks = (snapshot) => {
  if (!snapshot.context.selection)
    return !1;
  const startPoint = snapshot.context.selection.backward ? snapshot.context.selection.focus : snapshot.context.selection.anchor, endPoint = snapshot.context.selection.backward ? snapshot.context.selection.anchor : snapshot.context.selection.focus, startBlock = getSelectionStartBlock(snapshot), endBlock = getSelectionEndBlock(snapshot);
  if (!startBlock || !endBlock)
    return !1;
  const startBlockStartPoint = util_getTextBlockText.getBlockStartPoint({
    context: snapshot.context,
    block: startBlock
  }), endBlockEndPoint = util_isEmptyTextBlock.getBlockEndPoint({
    context: snapshot.context,
    block: endBlock
  });
  return util_isEmptyTextBlock.isEqualSelectionPoints(startBlockStartPoint, startPoint) && util_isEmptyTextBlock.isEqualSelectionPoints(endBlockEndPoint, endPoint);
};
function isBlockPath(path) {
  const firstSegment = path.at(0);
  return path.length === 1 && firstSegment !== void 0 && isRecord(firstSegment) && "_key" in firstSegment && typeof firstSegment._key == "string";
}
function isRecord(value) {
  return !!value && (typeof value == "object" || typeof value == "function");
}
function isSelectionExpanded(selection) {
  return selection ? !util_isEmptyTextBlock.isSelectionCollapsed(selection) : !1;
}
const getNextSpan = (snapshot) => {
  const selectionEndBlock = getSelectionEndBlock(snapshot), selectionEndPoint = getSelectionEndPoint(snapshot);
  if (!selectionEndBlock || !selectionEndPoint || !schema.isTextBlock(snapshot.context, selectionEndBlock.node))
    return;
  const selectionEndPointChildKey = util_getTextBlockText.getChildKeyFromSelectionPoint(selectionEndPoint);
  let endPointChildFound = !1, nextSpan;
  for (const child of selectionEndBlock.node.children) {
    if (child._key === selectionEndPointChildKey) {
      endPointChildFound = !0;
      continue;
    }
    if (schema.isSpan(snapshot.context, child) && endPointChildFound) {
      nextSpan = {
        node: child,
        path: [...selectionEndBlock.path, "children", {
          _key: child._key
        }]
      };
      break;
    }
  }
  return nextSpan;
}, getPreviousSpan = (snapshot) => {
  const selectionStartBlock = getSelectionStartBlock(snapshot), selectionStartPoint = selector_getSelectionText.getSelectionStartPoint(snapshot);
  if (!selectionStartBlock || !selectionStartPoint || !schema.isTextBlock(snapshot.context, selectionStartBlock.node))
    return;
  const selectionStartPointChildKey = util_getTextBlockText.getChildKeyFromSelectionPoint(selectionStartPoint);
  let previousSpan;
  for (const child of selectionStartBlock.node.children) {
    if (child._key === selectionStartPointChildKey)
      break;
    schema.isSpan(snapshot.context, child) && (previousSpan = {
      node: child,
      path: [...selectionStartBlock.path, "children", {
        _key: child._key
      }]
    });
  }
  return previousSpan;
}, getSelectedSpans = (snapshot) => {
  if (!snapshot.context.selection)
    return [];
  const selectedSpans = [], startPoint = selector_getSelectionText.getSelectionStartPoint(snapshot), endPoint = getSelectionEndPoint(snapshot);
  if (!startPoint || !endPoint)
    return selectedSpans;
  const startBlockKey = util_getTextBlockText.getBlockKeyFromSelectionPoint(startPoint), endBlockKey = util_getTextBlockText.getBlockKeyFromSelectionPoint(endPoint), startSpanKey = util_getTextBlockText.getChildKeyFromSelectionPoint(startPoint), endSpanKey = util_getTextBlockText.getChildKeyFromSelectionPoint(endPoint);
  if (!startBlockKey || !endBlockKey)
    return selectedSpans;
  const startBlockIndex = snapshot.blockIndexMap.get(startBlockKey), endBlockIndex = snapshot.blockIndexMap.get(endBlockKey);
  if (startBlockIndex === void 0 || endBlockIndex === void 0)
    return selectedSpans;
  const slicedValue = snapshot.context.value.slice(startBlockIndex, endBlockIndex + 1);
  let startBlockFound = !1;
  for (const block of slicedValue)
    if (block._key === startBlockKey && (startBlockFound = !0), !!schema.isTextBlock(snapshot.context, block)) {
      if (block._key === startBlockKey) {
        for (const child of block.children)
          if (schema.isSpan(snapshot.context, child)) {
            if (startSpanKey && child._key === startSpanKey) {
              if (startPoint.offset < child.text.length && selectedSpans.push({
                node: child,
                path: [{
                  _key: block._key
                }, "children", {
                  _key: child._key
                }]
              }), startSpanKey === endSpanKey)
                break;
              continue;
            }
            if (endSpanKey && child._key === endSpanKey) {
              endPoint.offset > 0 && selectedSpans.push({
                node: child,
                path: [{
                  _key: block._key
                }, "children", {
                  _key: child._key
                }]
              });
              break;
            }
            selectedSpans.length > 0 && selectedSpans.push({
              node: child,
              path: [{
                _key: block._key
              }, "children", {
                _key: child._key
              }]
            });
          }
        if (startBlockKey === endBlockKey)
          break;
        continue;
      }
      if (block._key === endBlockKey) {
        for (const child of block.children)
          if (schema.isSpan(snapshot.context, child)) {
            if (endSpanKey && child._key === endSpanKey) {
              endPoint.offset > 0 && selectedSpans.push({
                node: child,
                path: [{
                  _key: block._key
                }, "children", {
                  _key: child._key
                }]
              });
              break;
            }
            selectedSpans.push({
              node: child,
              path: [{
                _key: block._key
              }, "children", {
                _key: child._key
              }]
            });
          }
        break;
      }
      if (startBlockFound)
        for (const child of block.children)
          schema.isSpan(snapshot.context, child) && selectedSpans.push({
            node: child,
            path: [{
              _key: block._key
            }, "children", {
              _key: child._key
            }]
          });
    }
  return selectedSpans;
}, getMarkState = (snapshot) => {
  if (!snapshot.context.selection)
    return;
  let selection = snapshot.context.selection;
  if (!selector_getSelectionText.getFocusTextBlock(snapshot))
    return;
  if (isBlockPath(selection.anchor.path)) {
    const spanSelectionPoint = util_getTextBlockText.blockOffsetToSpanSelectionPoint({
      context: snapshot.context,
      blockOffset: {
        path: selection.anchor.path,
        offset: selection.anchor.offset
      },
      direction: selection.backward ? "backward" : "forward"
    });
    selection = spanSelectionPoint ? {
      ...selection,
      anchor: spanSelectionPoint
    } : selection;
  }
  if (isBlockPath(selection.focus.path)) {
    const spanSelectionPoint = util_getTextBlockText.blockOffsetToSpanSelectionPoint({
      context: snapshot.context,
      blockOffset: {
        path: selection.focus.path,
        offset: selection.focus.offset
      },
      direction: selection.backward ? "backward" : "forward"
    });
    selection = spanSelectionPoint ? {
      ...selection,
      focus: spanSelectionPoint
    } : selection;
  }
  const focusSpan = selector_getSelectionText.getFocusSpan({
    ...snapshot,
    context: {
      ...snapshot.context,
      selection
    }
  });
  if (!focusSpan)
    return;
  if (isSelectionExpanded(selection)) {
    const selectedSpans = getSelectedSpans({
      ...snapshot,
      context: {
        ...snapshot.context,
        selection
      }
    });
    let index = 0, marks2 = [];
    for (const span of selectedSpans) {
      if (index === 0)
        marks2 = span.node.marks ?? [];
      else {
        if (span.node.marks?.length === 0) {
          marks2 = [];
          continue;
        }
        marks2 = marks2.filter((mark) => (span.node.marks ?? []).some((spanMark) => spanMark === mark));
      }
      index++;
    }
    return {
      state: "unchanged",
      marks: marks2
    };
  }
  const decorators = snapshot.context.schema.decorators.map((decorator) => decorator.name), marks = focusSpan.node.marks ?? [], marksWithoutAnnotations = marks.filter((mark) => decorators.includes(mark)), spanHasAnnotations = marks.length > marksWithoutAnnotations.length, spanIsEmpty = focusSpan.node.text.length === 0, atTheBeginningOfSpan = snapshot.context.selection.anchor.offset === 0, atTheEndOfSpan = snapshot.context.selection.anchor.offset === focusSpan.node.text.length, previousSpan = getPreviousSpan({
    ...snapshot,
    context: {
      ...snapshot.context,
      selection
    }
  }), nextSpan = getNextSpan({
    ...snapshot,
    context: {
      ...snapshot.context,
      selection
    }
  }), nextSpanAnnotations = nextSpan?.node?.marks?.filter((mark) => !decorators.includes(mark)) ?? [], spanAnnotations = marks.filter((mark) => !decorators.includes(mark)), previousSpanHasAnnotations = previousSpan ? previousSpan.node.marks?.some((mark) => !decorators.includes(mark)) : !1, previousSpanHasSameAnnotations = previousSpan ? previousSpan.node.marks?.filter((mark) => !decorators.includes(mark)).every((mark) => marks.includes(mark)) : !1, previousSpanHasSameAnnotation = previousSpan ? previousSpan.node.marks?.some((mark) => !decorators.includes(mark) && marks.includes(mark)) : !1, previousSpanHasSameMarks = previousSpan ? previousSpan.node.marks?.every((mark) => marks.includes(mark)) : !1, nextSpanSharesSomeAnnotations = spanAnnotations.some((mark) => nextSpanAnnotations?.includes(mark));
  if (spanHasAnnotations && !spanIsEmpty) {
    if (atTheBeginningOfSpan) {
      if (previousSpanHasSameMarks)
        return {
          state: "changed",
          previousMarks: marks,
          marks: previousSpan?.node.marks ?? []
        };
      if (previousSpanHasSameAnnotations)
        return {
          state: "changed",
          previousMarks: marks,
          marks: previousSpan?.node.marks ?? []
        };
      if (previousSpanHasSameAnnotation)
        return {
          state: "unchanged",
          previousMarks: marks,
          marks: focusSpan.node.marks ?? []
        };
      if (!previousSpan)
        return {
          state: "changed",
          previousMarks: marks,
          marks: []
        };
    }
    if (atTheEndOfSpan) {
      if (nextSpan && nextSpanSharesSomeAnnotations && nextSpanAnnotations.length < spanAnnotations.length || !nextSpanSharesSomeAnnotations)
        return {
          state: "changed",
          previousMarks: marks,
          marks: nextSpan?.node.marks ?? []
        };
      if (!nextSpan)
        return {
          state: "changed",
          previousMarks: marks,
          marks: []
        };
    }
  }
  return atTheBeginningOfSpan && !spanIsEmpty && previousSpan ? previousSpanHasAnnotations ? {
    state: "changed",
    marks,
    previousMarks: previousSpan?.node.marks ?? []
  } : {
    state: "changed",
    previousMarks: marks,
    marks: (previousSpan?.node.marks ?? []).filter((mark) => decorators.includes(mark))
  } : {
    state: "unchanged",
    marks
  };
};
function getActiveDecorators(snapshot) {
  const schema2 = snapshot.context.schema, decoratorState = snapshot.decoratorState, markState = getMarkState(snapshot), decorators = schema2.decorators.map((decorator) => decorator.name);
  let activeDecorators = (markState?.marks ?? []).filter((mark) => decorators.includes(mark));
  for (const decorator in decoratorState)
    decoratorState[decorator] === !1 ? activeDecorators = activeDecorators.filter((activeDecorator) => activeDecorator !== decorator) : decoratorState[decorator] === !0 && (activeDecorators.includes(decorator) || activeDecorators.push(decorator));
  return activeDecorators;
}
const getTrimmedSelection = (snapshot) => {
  if (!snapshot.context.selection)
    return snapshot.context.selection;
  const startPoint = util_getTextBlockText.getSelectionStartPoint(snapshot.context.selection), endPoint = util_getTextBlockText.getSelectionEndPoint(snapshot.context.selection), startBlockKey = util_getTextBlockText.getBlockKeyFromSelectionPoint(startPoint), startChildKey = util_getTextBlockText.getChildKeyFromSelectionPoint(startPoint), endBlockKey = util_getTextBlockText.getBlockKeyFromSelectionPoint(endPoint), endChildKey = util_getTextBlockText.getChildKeyFromSelectionPoint(endPoint);
  if (!startBlockKey || !endBlockKey)
    return snapshot.context.selection;
  const startBlockIndex = snapshot.blockIndexMap.get(startBlockKey), endBlockIndex = snapshot.blockIndexMap.get(endBlockKey);
  if (startBlockIndex === void 0 || endBlockIndex === void 0)
    return snapshot.context.selection;
  const slicedValue = snapshot.context.value.slice(startBlockIndex, endBlockIndex + 1);
  let startBlockFound = !1, adjustedStartPoint, trimStartPoint = !1, adjustedEndPoint, trimEndPoint = !1, previousPotentialEndpoint;
  for (const block of slicedValue)
    if (!(block._key === startBlockKey && (startBlockFound = !0, schema.isTextBlock(snapshot.context, block) && util_isEmptyTextBlock.isEmptyTextBlock(snapshot.context, block))) && startBlockFound && schema.isTextBlock(snapshot.context, block)) {
      if (block._key === endBlockKey && util_isEmptyTextBlock.isEmptyTextBlock(snapshot.context, block))
        break;
      for (const child of block.children) {
        if (child._key === endChildKey && (!schema.isSpan(snapshot.context, child) || endPoint.offset === 0)) {
          adjustedEndPoint = previousPotentialEndpoint ? {
            path: [{
              _key: previousPotentialEndpoint.blockKey
            }, "children", {
              _key: previousPotentialEndpoint.span._key
            }],
            offset: previousPotentialEndpoint.span.text.length
          } : void 0, trimEndPoint = !0;
          break;
        }
        if (trimStartPoint) {
          const lonelySpan = schema.isSpan(snapshot.context, child) && block.children.length === 1;
          (schema.isSpan(snapshot.context, child) && child.text.length > 0 || lonelySpan) && (adjustedStartPoint = {
            path: [{
              _key: block._key
            }, "children", {
              _key: child._key
            }],
            offset: 0
          }, previousPotentialEndpoint = {
            blockKey: block._key,
            span: child
          }, trimStartPoint = !1);
          continue;
        }
        if (child._key === startChildKey) {
          if (!schema.isSpan(snapshot.context, child)) {
            trimStartPoint = !0;
            continue;
          }
          if (startPoint.offset === child.text.length) {
            trimStartPoint = !0, previousPotentialEndpoint = child.text.length > 0 ? {
              blockKey: block._key,
              span: child
            } : previousPotentialEndpoint;
            continue;
          }
        }
        previousPotentialEndpoint = schema.isSpan(snapshot.context, child) && child.text.length > 0 ? {
          blockKey: block._key,
          span: child
        } : previousPotentialEndpoint;
      }
      if (block._key === endBlockKey)
        break;
    }
  const trimmedSelection = snapshot.context.selection.backward ? {
    anchor: trimEndPoint && adjustedEndPoint ? adjustedEndPoint : endPoint,
    focus: adjustedStartPoint ?? startPoint,
    backward: !0
  } : {
    anchor: adjustedStartPoint ?? startPoint,
    focus: trimEndPoint && adjustedEndPoint ? adjustedEndPoint : endPoint
  };
  if (selector_getSelectionText.isSelectionCollapsed({
    context: {
      ...snapshot.context,
      selection: trimmedSelection
    }
  })) {
    const focusTextBlock = selector_getSelectionText.getFocusTextBlock({
      ...snapshot,
      context: {
        ...snapshot.context,
        selection: trimmedSelection
      }
    });
    if (focusTextBlock && !util_isEmptyTextBlock.isEmptyTextBlock(snapshot.context, focusTextBlock.node))
      return null;
  }
  return trimmedSelection;
}, getNextInlineObject = (snapshot) => {
  const focusTextBlock = selector_getSelectionText.getFocusTextBlock(snapshot), selectionEndPoint = getSelectionEndPoint(snapshot), selectionEndPointChildKey = selectionEndPoint && types.isKeySegment(selectionEndPoint.path[2]) ? selectionEndPoint.path[2]._key : void 0;
  if (!focusTextBlock || !selectionEndPointChildKey)
    return;
  let endPointChildFound = !1, inlineObject;
  for (const child of focusTextBlock.node.children) {
    if (child._key === selectionEndPointChildKey) {
      endPointChildFound = !0;
      continue;
    }
    if (!schema.isSpan(snapshot.context, child) && endPointChildFound) {
      inlineObject = {
        node: child,
        path: [...focusTextBlock.path, "children", {
          _key: child._key
        }]
      };
      break;
    }
  }
  return inlineObject;
}, getCaretWordSelection = (snapshot) => {
  if (!snapshot.context.selection || !selector_getSelectionText.isSelectionCollapsed(snapshot))
    return null;
  const focusTextBlock = selector_getSelectionText.getFocusTextBlock(snapshot), selectionStartPoint = selector_getSelectionText.getSelectionStartPoint(snapshot), selectionStartOffset = selectionStartPoint ? util_getTextBlockText.spanSelectionPointToBlockOffset({
    context: snapshot.context,
    selectionPoint: selectionStartPoint
  }) : void 0;
  if (!focusTextBlock || !selectionStartPoint || !selectionStartOffset)
    return null;
  const previousInlineObject = selector_getSelectionText.getPreviousInlineObject(snapshot), blockStartPoint = util_getTextBlockText.getBlockStartPoint({
    context: snapshot.context,
    block: focusTextBlock
  }), textDirectlyBefore = selector_getSelectionText.getSelectionText({
    ...snapshot,
    context: {
      ...snapshot.context,
      selection: {
        anchor: previousInlineObject ? {
          path: previousInlineObject.path,
          offset: 0
        } : blockStartPoint,
        focus: selectionStartPoint
      }
    }
  }).split(/\s+/).at(-1), nextInlineObject = getNextInlineObject(snapshot), blockEndPoint = util_isEmptyTextBlock.getBlockEndPoint({
    context: snapshot.context,
    block: focusTextBlock
  }), textDirectlyAfter = selector_getSelectionText.getSelectionText({
    ...snapshot,
    context: {
      ...snapshot.context,
      selection: {
        anchor: selectionStartPoint,
        focus: nextInlineObject ? {
          path: nextInlineObject.path,
          offset: 0
        } : blockEndPoint
      }
    }
  }).split(/\s+/).at(0);
  if ((textDirectlyBefore === void 0 || textDirectlyBefore === "") && (textDirectlyAfter === void 0 || textDirectlyAfter === ""))
    return null;
  const caretWordStartOffset = textDirectlyBefore ? {
    ...selectionStartOffset,
    offset: selectionStartOffset.offset - textDirectlyBefore.length
  } : selectionStartOffset, caretWordEndOffset = textDirectlyAfter ? {
    ...selectionStartOffset,
    offset: selectionStartOffset.offset + textDirectlyAfter.length
  } : selectionStartOffset, caretWordStartSelectionPoint = util_getTextBlockText.blockOffsetToSpanSelectionPoint({
    context: snapshot.context,
    blockOffset: caretWordStartOffset,
    direction: "backward"
  }), caretWordEndSelectionPoint = util_getTextBlockText.blockOffsetToSpanSelectionPoint({
    context: snapshot.context,
    blockOffset: caretWordEndOffset,
    direction: "forward"
  });
  if (!caretWordStartSelectionPoint || !caretWordEndSelectionPoint)
    return null;
  const caretWordSelection = {
    anchor: caretWordStartSelectionPoint,
    focus: caretWordEndSelectionPoint
  };
  return selector_getSelectionText.isSelectionExpanded({
    context: {
      ...snapshot.context,
      selection: caretWordSelection
    }
  }) ? caretWordSelection : null;
};
function getActiveAnnotationsMarks(snapshot) {
  const schema2 = snapshot.context.schema;
  return (getMarkState(snapshot)?.marks ?? []).filter((mark) => !schema2.decorators.map((decorator) => decorator.name).includes(mark));
}
function isActiveAnnotation(annotation, options) {
  return (snapshot) => {
    if ((options?.mode ?? "full") === "partial")
      return selector_getSelectionText.getSelectedValue(snapshot).flatMap((block) => schema.isTextBlock(snapshot.context, block) ? block.markDefs ?? [] : []).some((markDef) => markDef._type === annotation);
    const selectionMarkDefs = getSelectedBlocks(snapshot).flatMap((block) => schema.isTextBlock(snapshot.context, block.node) ? block.node.markDefs ?? [] : []), activeAnnotations = getActiveAnnotationsMarks(snapshot);
    return selectionMarkDefs.filter((markDef) => markDef._type === annotation && activeAnnotations.includes(markDef._key)).length > 0;
  };
}
const getFocusBlockObject = (snapshot) => {
  const focusBlock = selector_getSelectionText.getFocusBlock(snapshot);
  return focusBlock && !schema.isTextBlock(snapshot.context, focusBlock.node) ? {
    node: focusBlock.node,
    path: focusBlock.path
  } : void 0;
}, getNextBlock = (snapshot) => {
  const selectionEndBlock = getSelectionEndBlock(snapshot);
  if (!selectionEndBlock)
    return;
  const index = snapshot.blockIndexMap.get(selectionEndBlock.node._key);
  if (index === void 0 || index === snapshot.context.value.length - 1)
    return;
  const nextBlock = snapshot.context.value.at(index + 1);
  return nextBlock ? {
    node: nextBlock,
    path: [{
      _key: nextBlock._key
    }]
  } : void 0;
}, getPreviousBlock = (snapshot) => {
  const selectionStartBlock = getSelectionStartBlock(snapshot);
  if (!selectionStartBlock)
    return;
  const index = snapshot.blockIndexMap.get(selectionStartBlock.node._key);
  if (index === void 0 || index === 0)
    return;
  const previousBlock = snapshot.context.value.at(index - 1);
  return previousBlock ? {
    node: previousBlock,
    path: [{
      _key: previousBlock._key
    }]
  } : void 0;
}, getFirstBlock = (snapshot) => {
  const node = snapshot.context.value[0];
  return node ? {
    node,
    path: [{
      _key: node._key
    }]
  } : void 0;
}, getLastBlock = (snapshot) => {
  const node = snapshot.context.value[snapshot.context.value.length - 1] ? snapshot.context.value[snapshot.context.value.length - 1] : void 0;
  return node ? {
    node,
    path: [{
      _key: node._key
    }]
  } : void 0;
};
function isAtTheEndOfBlock(block) {
  return (snapshot) => {
    if (!snapshot.context.selection || !selector_getSelectionText.isSelectionCollapsed(snapshot))
      return !1;
    const blockEndPoint = util_isEmptyTextBlock.getBlockEndPoint({
      context: snapshot.context,
      block
    });
    return util_isEmptyTextBlock.isEqualSelectionPoints(snapshot.context.selection.focus, blockEndPoint);
  };
}
function isAtTheStartOfBlock(block) {
  return (snapshot) => {
    if (!snapshot.context.selection || !selector_getSelectionText.isSelectionCollapsed(snapshot))
      return !1;
    const blockStartPoint = util_getTextBlockText.getBlockStartPoint({
      context: snapshot.context,
      block
    });
    return util_isEmptyTextBlock.isEqualSelectionPoints(snapshot.context.selection.focus, blockStartPoint);
  };
}
const getFocusListBlock = (snapshot) => {
  const focusTextBlock = selector_getSelectionText.getFocusTextBlock(snapshot);
  return focusTextBlock && util_getTextBlockText.isListBlock(snapshot.context, focusTextBlock.node) ? {
    node: focusTextBlock.node,
    path: focusTextBlock.path
  } : void 0;
};
function isActiveDecorator(decorator) {
  return (snapshot) => {
    if (selector_getSelectionText.isSelectionExpanded(snapshot)) {
      const selectedSpans = getSelectedSpans(snapshot);
      return selectedSpans.length > 0 && selectedSpans.every((span) => span.node.marks?.includes(decorator));
    }
    return getActiveDecorators(snapshot).includes(decorator);
  };
}
const getActiveAnnotations = (snapshot) => {
  if (!snapshot.context.selection)
    return [];
  const selectedBlocks = getSelectedBlocks(snapshot), activeAnnotations = (getMarkState(snapshot)?.marks ?? []).filter((mark) => !snapshot.context.schema.decorators.map((decorator) => decorator.name).includes(mark));
  return selectedBlocks.flatMap((block) => schema.isTextBlock(snapshot.context, block.node) ? block.node.markDefs ?? [] : []).filter((markDef) => activeAnnotations.includes(markDef._key));
}, getSelectedTextBlocks = (snapshot) => {
  if (!snapshot.context.selection)
    return [];
  const selectedTextBlocks = [], startPoint = util_getTextBlockText.getSelectionStartPoint(snapshot.context.selection), endPoint = util_getTextBlockText.getSelectionEndPoint(snapshot.context.selection), startBlockKey = util_getTextBlockText.getBlockKeyFromSelectionPoint(startPoint), endBlockKey = util_getTextBlockText.getBlockKeyFromSelectionPoint(endPoint);
  if (!startBlockKey || !endBlockKey)
    return selectedTextBlocks;
  const startBlockIndex = snapshot.blockIndexMap.get(startBlockKey), endBlockIndex = snapshot.blockIndexMap.get(endBlockKey);
  if (startBlockIndex === void 0 || endBlockIndex === void 0)
    return selectedTextBlocks;
  const slicedValue = snapshot.context.value.slice(startBlockIndex, endBlockIndex + 1);
  for (const block of slicedValue) {
    if (block._key === startBlockKey) {
      if (schema.isTextBlock(snapshot.context, block) && selectedTextBlocks.push({
        node: block,
        path: [{
          _key: block._key
        }]
      }), startBlockKey === endBlockKey)
        break;
      continue;
    }
    if (block._key === endBlockKey) {
      schema.isTextBlock(snapshot.context, block) && selectedTextBlocks.push({
        node: block,
        path: [{
          _key: block._key
        }]
      });
      break;
    }
    selectedTextBlocks.length > 0 && schema.isTextBlock(snapshot.context, block) && selectedTextBlocks.push({
      node: block,
      path: [{
        _key: block._key
      }]
    });
  }
  return selectedTextBlocks;
}, getActiveListItem = (snapshot) => {
  if (!snapshot.context.selection)
    return;
  const selectedTextBlocks = getSelectedBlocks(snapshot).map((block) => block.node).filter((block) => schema.isTextBlock(snapshot.context, block)), firstTextBlock = selectedTextBlocks.at(0);
  if (!firstTextBlock)
    return;
  const firstListItem = firstTextBlock.listItem;
  if (firstListItem && selectedTextBlocks.every((block) => block.listItem === firstListItem))
    return firstListItem;
};
function isActiveListItem(listItem) {
  return (snapshot) => getActiveListItem(snapshot) === listItem;
}
const getActiveStyle = (snapshot) => {
  if (!snapshot.context.selection)
    return;
  const selectedTextBlocks = getSelectedBlocks(snapshot).map((block) => block.node).filter((block) => schema.isTextBlock(snapshot.context, block)), firstTextBlock = selectedTextBlocks.at(0);
  if (!firstTextBlock)
    return;
  const firstStyle = firstTextBlock.style;
  if (firstStyle && selectedTextBlocks.every((block) => block.style === firstStyle))
    return firstStyle;
};
function isActiveStyle(style) {
  return (snapshot) => getActiveStyle(snapshot) === style;
}
exports.getActiveAnnotations = getActiveAnnotations;
exports.getActiveAnnotationsMarks = getActiveAnnotationsMarks;
exports.getActiveDecorators = getActiveDecorators;
exports.getActiveListItem = getActiveListItem;
exports.getActiveStyle = getActiveStyle;
exports.getCaretWordSelection = getCaretWordSelection;
exports.getFirstBlock = getFirstBlock;
exports.getFocusBlockObject = getFocusBlockObject;
exports.getFocusInlineObject = getFocusInlineObject;
exports.getFocusListBlock = getFocusListBlock;
exports.getLastBlock = getLastBlock;
exports.getMarkState = getMarkState;
exports.getNextBlock = getNextBlock;
exports.getNextInlineObject = getNextInlineObject;
exports.getPreviousBlock = getPreviousBlock;
exports.getSelectedBlocks = getSelectedBlocks;
exports.getSelectedSpans = getSelectedSpans;
exports.getSelectedTextBlocks = getSelectedTextBlocks;
exports.getSelectionEndBlock = getSelectionEndBlock;
exports.getSelectionEndPoint = getSelectionEndPoint;
exports.getSelectionStartBlock = getSelectionStartBlock;
exports.getTrimmedSelection = getTrimmedSelection;
exports.isActiveAnnotation = isActiveAnnotation;
exports.isActiveDecorator = isActiveDecorator;
exports.isActiveListItem = isActiveListItem;
exports.isActiveStyle = isActiveStyle;
exports.isAtTheEndOfBlock = isAtTheEndOfBlock;
exports.isAtTheStartOfBlock = isAtTheStartOfBlock;
exports.isOverlappingSelection = isOverlappingSelection;
exports.isPointAfterSelection = isPointAfterSelection;
exports.isPointBeforeSelection = isPointBeforeSelection;
exports.isSelectingEntireBlocks = isSelectingEntireBlocks;
