// @portabletext/editor@2.14.2/lib/_chunks-es/util.is-empty-text-block.js
import { isTextBlock, isSpan } from "@portabletext/schema";
import { getTextBlockText } from "./util.get-text-block-text.js";
function getBlockEndPoint({
  context,
  block
}) {
  if (isTextBlock(context, block.node)) {
    const lastChild = block.node.children[block.node.children.length - 1];
    if (lastChild)
      return {
        path: [...block.path, "children", {
          _key: lastChild._key
        }],
        offset: isSpan(context, lastChild) ? lastChild.text.length : 0
      };
  }
  return {
    path: block.path,
    offset: 0
  };
}
function isSelectionCollapsed(selection) {
  return selection ? JSON.stringify(selection.anchor.path) === JSON.stringify(selection.focus.path) && selection.anchor.offset === selection.focus.offset : !1;
}
function isEqualSelectionPoints(a, b) {
  return a.offset === b.offset && JSON.stringify(a.path) === JSON.stringify(b.path);
}
function isEmptyTextBlock(context, block) {
  if (!isTextBlock(context, block))
    return !1;
  const onlyText = block.children.every((child) => isSpan(context, child)), blockText = getTextBlockText(block);
  return onlyText && blockText === "";
}
export {
  getBlockEndPoint,
  isEmptyTextBlock,
  isEqualSelectionPoints,
  isSelectionCollapsed
};
